File: euler/tasks/79/main.py
import sys
import collections

deps = collections.defaultdict(list)

with open(sys.argv[1], 'r') as f:
    for l in f:
        l = l.strip()

        if l:
            deps[l[0]].append(l[1])
            deps[l[1]].append(l[2])

v = set()

def visit(n):
    for d in deps.get(n, []):
        yield from visit(d)

    if n not in v:
        v.add(n)

        yield n

print(''.join([str(x) for x in reversed(list(visit('7')))]))

File: euler/tasks/79/main.cpp
#include <euler/lib/euler.h>

int main() {
    std::map<char, std::vector<char>> deps;

    for (auto& l : read_lines("./79_keylog.txt")) {
        deps[l[0]].push_back(l[1]);
        deps[l[1]].push_back(l[2]);
    }

    std::set<char> v;
    std::string res;
    std::function<void (char)> visit;

    visit = [&] (char n) {
        for (auto d : deps[n]) {
            visit(d);
        }

        if (v.find(n) == v.end()) {
            v.insert(n);
            res += n;
        }
    };

    visit('7');

    std::cout << reversed(res) << std::endl;
}
